Sources/SwiftFusion/Inference/NonlinearFactorGraph.swift

```diff
@@ -0,0 +1,49 @@
+// Copyright 2019 The SwiftFusion Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+import TensorFlow
+
+/// A factor graph for nonlinear problems
+/// TODO(fan): Add noise model
+public struct NonlinearFactorGraph: FactorGraph {
+  public typealias KeysType = Array<Int>
+  
+  public typealias FactorsType = Array<AnyNonlinearFactor>
```

Comment:
You said during our Friday meeting that you still haven't been able to eliminate `AnyNonlinearFactor` and replace this with `Array<NonlinearFactor>`.

I figured out why. The `associatedtype FactorsType : Collection where FactorsType.Element: Factor` in `protocol FactorGraph` doesn't work when you use `Array<NonlinearFactor>`, and the reason it doesn't work is that existentials can't conform to protocols. (This is a limitation of the Swift type system that probably won't go away any time soon.)

The easiest way to solve this would be to delete the `protocol FactorGraph`. Nothing actually relies on abstracting over different `FactorGraph` implementations yet, so this doesn't break anything.

I think this is worth it because deleting `AnyNonlinearFactor` makes everything a lot simpler and nicer.

It's also probably possible to redesign the `FactorGraph` protocol so that it doesn't have a constraint that breaks `Array<NonlinearFactor>`. I'd suggest writing the `FactorGraph` protocol later when we have a use case for that, because the use case will make it clearer what `protocol FactorGraph` needs to have.

Reply:
@marcrasi Oh that's a great find! I'll remove the protocol for now :)

Thanks for the explanation!